설정 화면에 추가된 UI 구현, 사용자 설정 값 관리하는 UseCase와 Repository 구현

sanghyeok-kim opened the pull request:
### 🔖 관련 이슈
- #43 

<br> 

### ⚒️ 작업 내역

- [x] 프로필 정보 수정 화면 UI 구현
- [x] 프로필 정보 수정 UseCase 구현
- [x] 로그인 정보, 자동 로그인 여부, 버전 정보에 대한 UseCase 구현
- [x] 탈퇴 화면과 의견 보내기 화면에 사용될 커스텀 TextView 구현
- [x] 의견 보내기 화면 UI 구현
- [ ] 의견 보내기 UseCase 구현
- [ ] 로그아웃 UseCase 구현
- [ ] 탈퇴 UseCase 구현

<br>

### 📝 부가 설명

- 설정 화면에서 사용되는 로직에 대한 UseCase들을 UserDefaults를 사용하여 구현했습니다. 
 SharedUseCase 모듈에 각각 구현했으니 필요한 경우 해당 모듈을 import하여 사용하면 됩니다.
  - LoginType(로그인 정보)
  - Nickname(사용자 닉네임)
  - FeedVisibility(공개범위 설정)
  - AutoLogin(자동 로그인 사용)
  - ProfileCharacter(프로필 이미지)


- 화면 전환 과정에서 '부모 VC의 VM'-'전환되는 VC의 VM'간 데이터 동기화가 필요해서,
  SettingCoordinator 모듈에 RxSwift 의존성을 추가했습니다.  
  추후 다른 Coordinator 모듈도 필요시 동일하게 RxSwift 의존성을 추가해서 사용하면 됩니다.


- 회의때 이야기 했던 재사용될 수 있는 뷰 관련 요소들은 우선 중복을 고려하지 않고 일단 구현해두긴 했습니다만  
이후에 별도의 모듈로 분리하는 등 다른 방안을 찾아봐야할 것 같습니다.
  - DesignSystem + Rx 익스텐션
  - CoreEntity와 DesignSystem Asset간 Mapping 

- 코드 사이닝 관련 설정과 gitignore 수정 등 공통으로 적용될 변경 사항은 별도의 이슈로 분리했습니다.
  #48 

<br> 

### 📑 첨부 자료

- UseCase 사용하여 UserDefaults에 저장

https://github.com/LifePoop/LifePoop_iOS/assets/57667738/43faf2f5-4095-4266-bbff-d44fefe48b8e

- LifePopoTextView (키보드에 가릴 경우 스크롤뷰 contentOffest 변경, 텍스트 길이 제한)

https://github.com/LifePoop/LifePoop_iOS/assets/57667738/b16fe81d-8fe1-49c3-b015-223983e07633

sanghyeok-kim committed:
#43 feat: expand_down 이미지 에셋 추가

sanghyeok-kim committed:
#43 feat: 커스텀 Rx+viewWillAppear 구현

sanghyeok-kim committed:
#43 refactor: LifePoopButton의 backgroundColor 변경을 Configuration 방식에서 기존 방식으로 변경

- isHighlighted에 대한 backgroundColor 설정은 Configuration으로 잘 적용되지만, isEnabled의 경우 적용되지 않아 기존 방식(Configuration을 사용하지 않는 방식)으로 변경함

sanghyeok-kim committed:
#43 refactor: PaddingButton의 Padding Inset이 더 정확하게 설정되도록 변경

- 기존의 경우 기본적으로 제공되는 tite과 button간 Inset이 이미 적용되어있는 상태여서, 설정한 Padding보다 더 넓은 간격이 설정되는 문제가 있었음
- 이제 설정한 Padding(UIEdgeInsets)값 만큼 정확하게 Inset이 설정되도록 변경됨.

sanghyeok-kim committed:
#43 refactor: ColorSet을 수정된 에셋으로 교체

sanghyeok-kim committed:
#43 feat: Profile Image 에셋 추가

sanghyeok-kim committed:
#43 move: LoginType, FeedVisibility 열거형을 Utils 모듈에서 CoreEntity 모듈로 이동

sanghyeok-kim committed:
#43 feat: FeedbackViewController에 사용되는 SuggestionButton 컴포넌트 생성 및 구현

sanghyeok-kim committed:
#43 feat: FeedbackViewController, WithdrawViewController에서 사용되는 LifePoopTextView 생성 및 구현

sanghyeok-kim committed:
#43 rename: 변경된 에셋 이미지명으로 수정

sanghyeok-kim committed:
#43 refactor: NavigationBar BackButtonIndicator를 빈 이미지로 설정시 발생하는 레이아웃 경고 처리를 위해 이미지 크기 설정

sanghyeok-kim committed:
#43 feat: HomeCoordinator FinishFlow 구현

sanghyeok-kim committed:
#43 feat: SettingCoordinator FinishFlow 구현

sanghyeok-kim committed:
#43 move: SelectableTextRadioButtonView 디렉토리 위치 이동 (DesginSystem/UIButton -> DesginSystem/UIView)

sanghyeok-kim committed:
#43 feat: SelectableTextRadioButtonView를 상속받는 FeedVisibilitySelectionButtonView 생성 및 구현

sanghyeok-kim committed:
#43 feat: 설정 화면에 FeedVisibility 탭 추가 및 구현

sanghyeok-kim committed:
#43 refactor: WithdrawViewController의 textView 편집시 키보드에 가리지 않도록 수정

- UIScollView 추가, 키보드 오르내릴 때 마다 가려지는 영역의 높이 구한 뒤 contentOffset 수정

sanghyeok-kim committed:
#43 feat: CaseIterable의 인덱싱 기능을 위한 extension 구현

sanghyeok-kim committed:
#43 refactor: Utils 번들에 대한 접근을 Bundle extension으로 분리

sanghyeok-kim committed:
#43 feat: SwitchViewCell에 ViewModel 바인딩 로직 추가

sanghyeok-kim committed:
#43 refactor: SettingTapActionCellViewModel이 additionalText를 생성자로 직접 받지 않고 별도의 함수로 바인딩하도록 변경

sanghyeok-kim committed:
#43 feat: SharedDIContainer 생성 및 구현

sanghyeok-kim committed:
#43 remove: 불필요한 URLDataRepository, URLDataService 삭제

sanghyeok-kim committed:
#43 feat: 사용자 설정 관련 UserDefault UseCase, Repository 구현

sanghyeok-kim committed:
#43 feat: FeedbackViewModel, ViewController 바텀시트 제외하고 구현

sanghyeok-kim committed:
#43 feat: ProfileCharacter Entity 생성 및 구현

sanghyeok-kim committed:
#43 feat: Profile, ProfileEdit ViewController & ViewModel 생성 및 구현

sanghyeok-kim committed:
#43 feat: SelectableColor, SelectableStiffness Entity를 UserDefaults에 저장할 수 있도록 Codable 채택

sanghyeok-kim committed:
#43 feat: SettingCoordinator 모듈에 RxSwift, CoreEntity 모듈 의존성 추가하여 설정 화면에 대한 Coordinator 흐름 구현

sanghyeok-kim committed:
#43 feat: AppCoordinator가 HomeCoordinator의 completionDelegate를 채택할 수 있도록 HomeCoordinatorInterface 모듈 의존성 추가

sanghyeok-kim committed:
#43 feat: FeatureSettingUseCase에 SharedUseCase 모듈 의존성 추가

sanghyeok-kim committed:
#43 feat: SettingViewModel 세부적인 로직 추가 구현

- 임시 객체를 직접 생성해서 사용하던 방식에서 UseCase 사용하는 방식으로 변경
- FeedVisibility에 대한 UseCase와 Coordinator 흐름 추가

sanghyeok-kim committed:
#43 chore: Lint Rule에서 함수 길이 제한 규칙 삭제

sanghyeok-kim committed:
#43 프로필 설정 화면에서 사용할 profileImageEditView 커스텀 뷰 구현

sanghyeok-kim committed:
#43 refactor: LifePoopTextView 키보드 툴바 버튼 title이 한글로 표기되도록 변경 (done -> 완료)

junu0516 commented on Projects/App/Sources/SceneDelegate.swift:
```diff
@@ -48,18 +54,28 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
     func registerCoreDependencies() {
         CoreDIContainer.shared.register(service: AnyDataMapper.self) { AnyDataMapper(CoreExampleDataMapper()) }
         CoreDIContainer.shared.register(service: EndpointService.self) { URLSessionEndpointService.shared }
-        CoreDIContainer.shared.register(service: URLDataService.self) { URLSessionURLDataService.shared }
```
`URLSessionDataService` 는 원래 어떤 용도로 있던건가요??

sanghyeok-kim replied:
서버에서 이미지 URL을 받아와, 해당 URL의 이미지를 Data 타입으로 변환해주는 Service입니다.
초기에 Example로 구현했던 것들 중 하나인데, 우리 프로젝트에는 쓰이지 않을 기능이라 삭제했습니다.

junu0516 replied:
아하 확인 감사합니다 ㅎㅎ 코드 관련 히스토리(?) 확인차 여쭤봤었어요!

junu0516 commented on Projects/App/Sources/SceneDelegate.swift:
```diff
@@ -69,13 +85,17 @@ private extension SceneDelegate {
         HomeDIContainer.shared.register(service: HomeUseCase.self) { DefaultHomeUseCase() }
         HomeDIContainer.shared.register(service: HomeRepository.self) { DefaultHomeRepository() }
     }
+    
+    func registerSettingDependencies() {
+        SettingDIContainer.shared.register(service: UserSettingUseCase.self) { DefaultUserSettingUseCase() }
+    }
 }
 
 // MARK: - UI Setup
 
 private extension SceneDelegate {
     func configureNavigationBarBackButtonItem() {
-        let emptyImage = UIImage()
+        let emptyImage = UIGraphicsImageRenderer(size: CGSize(width: 1, height: 1)).image { _ in }
```
`UIGraphicsImageRender` 사용은 좀 더 자원을 효율적으로 사용하기 위함인가요? 목적이 궁금해서 질문 남겨습니다!

sanghyeok-kim replied:
피그마 요구사항대로 백버튼을 구현하기 위해서는
네비게이션바에 기본으로 제공되는 backIndicatorImage(백버튼에 기본으로 나타나는 '<'모양 이미지)를 없애야했는데,
처음엔 저렇게 단순히 빈 이미지 `let emptyImage = UIImage()`로 덮어써서 없애는 방식을 사용했습니다.

다만 그렇게하니까 emptyImage의 크기를 계산할 수 없어 레이아웃 경고가 발생했고,
이를 없애기 위해 크기를 직접 설정한 이미지를 할당하는 방식으로 변경한 코드입니다.
(1x1 크기로 설정한 것에는 특별한 이유가 없고, 레이아웃 경고를 없애기 위해 임의로 설정한 값이라 생각하시면 될거같습니다.)

저 방법 외에 back버튼을 커스터마이징할 수 있는 다른 방법이 있는지 찾아보았으나
결정적으로 backIndicatorImage를 없애는 메소드나 프로퍼티는 별도로 제공하지 않는걸로 파악되어서..ㅠㅠ
우선은 이 방법이 현재로서 제가 생각하는 Best Practice라서 저렇게 구현해두었습니다

(사실 저 레이아웃 경고는 뷰에 큰 영향을 끼치지 않아서 무시해도 되는 부분이긴한데,
뷰 디버깅 할 때 마다 신경이 쓰여서 없애버리려고 했습니다 ㅎㅎ)

junu0516 replied:
아하 그렇군요.. 궁금해서 질문드렸어용 덕분에 저도 배워갑니다 ㅎㅎ

junu0516 commented on Projects/Core/CoreEntity/Sources/LoginType.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+public enum LoginType: Codable {
+    case apple
+    case kakao
+}
```
아.. 이거 CoreEntity로 옮겨졌군요
바로 밑에 올린 PR에서 이것 때문에 CoreEntity에서 Utils 모듈 import 가능하게 해뒀는데, 해당 PR 머지되고 나면 이 부분 다시 원복시킬께요
(혹시 이전에 옮긴다고 미리 말해주신거면 제가 깜빡한듯해요 🙏)

junu0516 commented on Projects/Core/CoreEntity/Sources/SelectableStiffness.swift:
```diff
@@ -8,7 +8,7 @@
 import Foundation
 
-public enum SelectableStiffness: CaseIterable {
+public enum SelectableStiffness: Codable, CaseIterable { // FIXME: 네이밍 수정
     
     case soft
     case normal
```
Stiffness는 '강직도'를 나타내기 위해 사용한건데, 단순히 '딱딱한 정도'의 의미만 나타내고자 하면 'Hardness' 같은 단어도 적합해보여요

sanghyeok-kim replied:
피그마 요구사항이 수정되면서 '모양'이라고 표기되어 있는데,
'강도', '강직도' 같은 느낌의 네이밍 보다는 '모양', '형태'를 나타내는 'Shape'로 하는건 어떨까요?

junu0516 replied:
아 다시 보니까 모양으로 되어있네요 말씀하신대로 Shape가 제일 좋아보여요 ㅎㅎ 이후에 저걸로 바꿔놓을께요

junu0516 commented on Projects/DesignSystem/Sources/Component/UIButton/PaddingButton.swift:
```diff
@@ -8,13 +8,15 @@
 import UIKit
 
-public class PaddingButton: UIButton {
+open class PaddingButton: UIButton {
```
PaddingButton이 open이 된건 DesignSystem외부에서도 해당 클래스를 상속받는 경우를 가정한건가요??

sanghyeok-kim replied:
PaddingButton을 상속받는 SuggestionButton을 구현하는 과정에서 open으로 변경된 부분인데요
당시엔 SuggestionButton이 DesignSystem에 속해있지 않아서 그렇게 했는데,
이제 다시 DesignSystem으로 이동시켰기 때문에 public으로 변경하면 될 것 같습니다.
확인해주셔서 감사해요 ㅎㅎ

junu0516 replied:
아하 ㅋㅋ 알겠습니다!

junu0516 commented on Projects/DesignSystem/Sources/Component/UIView/ProfileImageEditView.swift:
```diff
@@ -0,0 +1,55 @@
+import UIKit
+
+import SnapKit
+
+public final class ProfileImageEditView: UIView {
+    
```
해당 컴포넌트는 설정 화면 모듈 외에도, 다른 모듈에서 사용되는 경우를 가정하고 DesignSystem 내부에 있는건가요 혹시??
코드 보다가 컴포넌트들이 대부분 DesignSystem 안에 있길래 궁금해서 질문 남겼어요!

sanghyeok-kim replied:
사실 지금 당장은 다른 모듈에서 사용되진 않지만,
기왕 커스텀 뷰를 별도로 만드는 김에 DesignSystem에 기본 컴포넌트로 구현했습니다 ㅎㅎ
혹시 또 다른 모듈에서 사용될 일이 생기면 바로 가져다 쓸 수 있게 하면 좋을 것 같아서요

junu0516 replied:
아하 그렇군요 ㅋㅋ 구현해주신거 혹시 추후에 재사용할만한거 있으면 가져다 쓸께요 미리 감사합니다~!

junu0516 commented on Projects/DesignSystem/Sources/Component/UIView/SelectableTextRadioButtonView.swift:
```diff
@@ -0,0 +1,103 @@
+    
+    public lazy var containerButton: UIButton = {
+        let button = UIButton()
+        button.addAction(UIAction(handler: { [weak self] _ in
+            guard let self else { return }
+            self.isSelected.toggle()
+            self.toggleRadioButton(isSelected: self.isSelected)
+        }), for: .touchUpInside)
```
handler로 넘긴 클로저는 이름있는 별도의 함수로 빼서, 해당 버튼에 어떤 행위가 부여되었는 지를 좀 더 분명히 나타내면 좋을 것 같습니다!

junu0516 commented on Projects/DesignSystem/Sources/Component/UIView/SelectableTextRadioButtonView.swift:
```diff
@@ -0,0 +1,103 @@
+        return label
+    }()
+    
+    public override var intrinsicContentSize: CGSize {
+        let width = radioButtonImageView.intrinsicContentSize.width
+        + descriptionLabel.intrinsicContentSize.width
+        + imagePadding.offset
+        let height = radioButtonImageView.intrinsicContentSize.height
+        return CGSize(width: width, height: height)
+    }
```
이 부분도 width와 height 계산을 위해 쓰인 피연산자들의 의미를 좀 더 분명하게 드러내면 좋을 듯 합니다.
(ex. 클로저 안에 각 피연산자의 의미가 드러난 지역변수 선언)

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingCoordinator/Sources/DefaultSettingCoordinator.swift:
```diff
@@ -67,13 +76,27 @@ private extension DefaultSettingCoordinator {
+        let viewModel = ProfileViewModel(coordinator: self, userNickname: userNickname)
         viewController.bind(viewModel: viewModel)
         navigationController.pushViewController(viewController, animated: true)
     }
```
로그인 기능과, 설정 기능이 합쳐지면 해당 부분은 추후 닉네임을 포함한 회원 정보가 담긴 모델을 넘겨주는 쪽으로 바뀌면 될 것 같네요

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingCoordinator/Sources/DefaultSettingCoordinator.swift:
```diff
@@ -92,8 +115,16 @@ private extension DefaultSettingCoordinator {
+
+private extension DefaultSettingCoordinator {
+    func presentBottomSheetController(contentViewController: UIViewController, heightRatio: Double) {
+        // FIXME: 스크롤시 Navigation Bar 영역이 어두워지지 않음
+        guard let presentedViewController = navigationController.topViewController else { return }
+        let bottomSheetController = BottomSheetController(
+            bottomSheetHeight: presentedViewController.view.bounds.height * heightRatio
+        )
+        bottomSheetController.setBottomSheet(contentViewController: contentViewController)
+        bottomSheetController.showBottomSheet(toParent: presentedViewController)
```
이 부분은 고치는 중인데, 우선은 아래와 같은 방법으로 문제를 해결할 수 있긴 합니다. 
이전에 말로만 한 번 공유했는데, 잘 전달이 안된듯해서 우선 기록차 글로 한 번 남겨놓을께요!
(사실상 편법이라, 이거 없이 사용 가능하게 고치는 중이에요)

1. 현재 코디네이터가 사용중인 navigationController에 투명한 뷰컨트롤러를 임의로 띄움
```swift
typealias TransparentBackgroundViewController = UIViewController
func presentTransparentBackgroundView() {
    let backgroundViewController = TransparentBackgroundViewController()
    backgroundViewController.modalPresentationStyle = .overFullScreen
    navigationController.present(backgroundViewController, animated: false)
}
```
2. `bottomSheetController.showBottomSheet(toParent:)`의 파라미터로 presentedViewController 대신 투명한 뷰컨트롤러 인스턴스를 넘김
    - navigationController.topViewController 대신 navigationController.presentedViewController 사용
    - topViewController은 push된 뷰컨트롤러 대상이기 때문에, bottomSheetController이 띄워지더라도 UINavigationController 자체의 특성상 네비게이션 바를 감싸지 못합니다.

sanghyeok-kim replied:
topViewController와 presentedViewController의 차이가 헷갈렸는데 다시 확인해주셔서 감사해요
그럼 일단 이 부분은 그대로 두고 나중에 고치신 다음에 변경해볼게요

junu0516 replied:
네네 ㅋㅋ 최종 변경하면 다시 공유드릴께요

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/FeedVisibility/FeedVisibilityViewController.swift:
```diff
@@ -0,0 +1,79 @@
+
+public final class FeedVisibilityViewController: LifePoopViewController, ViewType {
+    
+    private let feedVisibilityButtons = FeedVisibility.allCases.enumerated().map { (index, element) in
+        let buttonView = FeedVisibilitySelectionButtonView(index: index, title: element.text)
+        buttonView.descriptionLabel.font = .systemFont(ofSize: 18, weight: .bold)
+        return buttonView
+    }
```
FeedVisibility의 모든 케이스에 대해서 enumerated() 한 번, map 한 번 이렇게 총 2번의 동일한 반복문을 돌고있는 셈인데,
개인적으로 이렇게 인덱스값이 필요한 allCases라면 차라리 열거형 내부에서 정렬된 형태의 allCases와, 각 케이스별 index를 알려주면 어떨까 하는 생각이 들어요!

제 코드도 보다보면 동일하게 enumerated를 한 번 호출하고, 이어서 map이나 filter을 호출하는 경우가 있는데 불필요한 반복문을 줄일 필요가 있지 않을까.. 하는 생각이 들어서 코멘트 남겨봤습니다. 혹시 어떻게 생각하시나요?

sanghyeok-kim replied:
이 문제는 제가 열거형 타입을 UserDefaults에 CRUD할 때 Data 타입으로 인코딩/디코딩해서 저장하도록 구현해서 발생한 것 같네요.

저도 사실 FeedVisibility가 Index를 갖도록 하려했는데
그러려면 FeedVisibility 열거형의 RawRepresentable 타입으로 Int를 명시해줘야하고,

```swift
enum FeedVisibility: Int { // RawRepresentable을 Int로 명시
    ...
    var index: Int {
        return rawValue
    }
}
```
이렇게 할 경우 UserDefaults에 열거형을 Encoding/Decoding해서 읽고 쓸 수가 없더라구요
```
// 런타임 에러
Attempt to set a non-property-list object CoreEntity.FeedVisibility.private as an NSUserDefaults/CFPreferences value for key FeedVisibility
```

그렇다고 RawRepresentable 타입을 Int로 명시하지 않으면서 내부적으로 Int타입의 Index 프로퍼티를 갖게 하자니,
switch-case문으로 모든 case마다 Index값을 0부터 일일이 지정해줘야하므로 확장성이 떨어지는 문제가 있었구요.

---

서론이 길었는데 위 내용이 제가 enumerated를 쓸 수 밖에 없었던 배경이었고,
결론은 UserDefaultsRepository에 제네릭 메소드를 2개로 나눠서
1. Codable을 채택한 객체(`<T: Codable>`)는 기존처럼 인코딩/디코딩해서 Data 형태로 저장하고,
2. RawRepresentable 타입을 채택한 객체(`<T: RawRepresentable>`)는 읽고 쓸 때 RawValue를 사용하도록

리팩토링해서 FeedVisibility를 비롯한 다른 열거형 타입의 Entity 객체들에 대한 읽기/쓰기와 인덱싱 작업을 개선하면 될 것 같습니다~!

junu0516 replied:
그렇군요 ㅋㅋ 사실 진짜 별거 아닌데, 혹여나 반복문의 수를 줄여볼 수 있을까 싶어서 여쭤봤어요.
언급하신대로면 ㅋㅋ 현재 상태가 제일 좋은 것 같아서.. 이대로 두는게 좋은거같습니다

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/Feedback/FeedbackViewController.swift:
```diff
@@ -41,6 +102,110 @@ public final class FeedbackViewController: LifePoopViewController, ViewType {
+// MARK: - Keyboard Notification Methods
+
+private extension FeedbackViewController {
+    func addNotificationCenterKeyboardObserver() {
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(keyboardWillShow(notification:)),
+            name: UIResponder.keyboardWillShowNotification,
+            object: nil
+        )
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(keyboardWillHide(notification:)),
+            name: UIResponder.keyboardWillHideNotification,
+            object: nil
+        )
+    }
+    
+    @objc func keyboardWillShow(notification: NSNotification) {
+        let keyboardFrameNotification = notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? NSValue
+        let keyboardHeight = keyboardFrameNotification?.cgRectValue.height ?? .zero
+        
+        let keyboardTopY = view.frame.maxY - keyboardHeight
+        let textViewBottomY = suggestionTextView.convert(suggestionTextView.bounds, to: view).maxY
+        
+        let offsetY = textViewBottomY - keyboardTopY
+        scrollView.contentOffset.y += offsetY
+    }
+    
+    @objc func keyboardWillHide(notification: NSNotification) {
+        scrollView.setContentOffset(.zero, animated: true)
```
keyboardWillShow 내부의 각 연산들에 대해, 피연산자들의 의미를 좀 더 분명히하면 가독성이 좋아질 것 같아요!
추가로 이 부분 RxSwift의 RxKeyBoard 사용해보시는건 혹시 어떻게 생각하시나요?
지금도 충분히 괜찮아보여서 불필요한 라이브러리 적용이라 생각하시면 넘겨도 좋은데, 라이브러리를 사용하면 바인딩 메소드 안에 키보드 동작도 포함시킬 수 있어서 코드가 좀 더 짧아지지 않을까 해서 제안드렸습니다!

sanghyeok-kim replied:
RxKeyboard는 생각을 못했는데 괜찮은 아이디어네요
코드 길이가 확실히 짧아질 것 같습니다 시간이 나면 한 번 시도해 볼게요

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/Profile/ProfileEditViewController.swift:
```diff
@@ -0,0 +1,112 @@
+        return button
+    }()
+    
+    public var viewModel: ProfileEditViewModel?
+    private let disposeBag = DisposeBag()
+    
```
이전 PR에서도 궁금했는데 혹시 disposeBag이 let으로 선언된 이유가 따로 있을까요!!?

sanghyeok-kim replied:
ViewController 인스턴스가 살아있는 동안에 disposeBag이 변경되지 않으므로 let으로 선언했습니다.

ViewCell 타입처럼 prepareForReuse될 때 마다 새로운 disposeBag으로 교체해줘야 하는 경우에는 var로 선언하지만,
지금 같은 경우엔 그러지 않으니 let으로 선언해도 괜찮지 않을까요?

junu0516 replied:
네 ㅋㅋ 사람마다 let, var을 쓰는게 상이해서 혹시 따로 어떤 기준을 두고 하시나 여쭤보고 혹여나 여기서도 통일할 부분이 있나 여쭤봤어요!
말씀하신대로 상황에 따라 let, var 유동적으로 쓰면 될 것 같습니다.

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/Profile/ProfileEditViewController.swift:
```diff
@@ -0,0 +1,112 @@
+    
+    // MARK: - UI Setup
+    
+    public override func configureUI() {
+        super.configureUI()
+        
+    }
```
이 부분은 아직 미구현인걸로 보면 될까요!!?

sanghyeok-kim replied:
네 아직 프로필 이미지 편집시 하단에서 나타나는 BottomSheet 내부 UI가 완성되지 않아서 비워뒀습니다.
주석이라도 달아둘걸 그랬네요 ㅋㅋ

junu0516 replied:
아아 이미 확인해주셔서 괜찮아요 ㅋㅋ 아직 구현중이신 것 같았는데 혹시나 해서 여쭤봤어요

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/Profile/ProfileEditViewModel.swift:
```diff
@@ -0,0 +1,87 @@
+    
+    public let input = Input()
+    public let output = Output()
+    public let state: State
```
`State` 타입을 아얘 ViewModel 공통 타입에 Input, Output과 함께 포함시키는건 어떠세요!?
저같은 경우에도 이전에 구현한 뷰모델들 보면 State에 해당하는 속성들이 좀 있어서, 혹시 괜찮다고 생각하시면 State도 같이 정의해놓고 필요에 따라 가져다 쓰면 괜찮을듯 싶어요!

다만 State가 필요하지 않은 뷰모델들이 꽤 있기 때문에 빈 내부 타입을 생성하는게 좀 비효율적이라 생각하시면 그냥 넘기는게 나을 것 같아요 ㅋㅋ

sanghyeok-kim replied:
저도 고민했던 부분인데 말씀하신대로 State가 필요한 ViewModel이 있는 반면 그렇지 않은 ViewModel도 많아서..
ViewModel 프로토콜에 넣으면, "이건 필수로 구현해야해"하는 느낌이 있는지라 굳이 State까지는 안넣어도 될 것 같아요

사실 프로젝트에 ReactorKit같이 ViewModel 바인딩 흐름을 정형화할 수 있는 라이브러리를 적용했다면
이 부분이 더 명확했을 것 같긴하네요 ㅎㅎ

junu0516 replied:
그렇군요 ㅋㅋ 확인 감사합니다 ㅎㅎㅎ

junu0516 commented on Projects/Utils/Sources/Bundle/BundleResourceAccessor.swift:
```diff
@@ -0,0 +1,20 @@
+
+import Foundation
+
+public struct BundleResourceAccessor {
+    public static var appVersion: String {
+        return Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? ""
+    }
+    
+    public static func readText(from file: TextFile) -> String? {
+        guard let path = Bundle.utils?.path(forResource: file.name, ofType: "txt") else { return nil }
+        return try? String(contentsOfFile: path, encoding: .utf8)
+    }
+}
```
Bundle에서 필요한 값을 읽어오지 못할 경우에는 빈 값보다는 예외처리하는게 괜찮을 것 같아요.
혹여나 런타임 시점에 Bundle에 있는 값을 제대로 못불러오는 경우가 발생하면, 좀 더 상황을 빨리 파악할 수 있을 것 같습니다!

sanghyeok-kim replied:
좋은 생각이네요~
안그래도 저번에 코드 사이닝 관련 작업 이후로 bundle identifier가 수정되는 바람에
기존에 작성해뒀던 Bundle.utils의 identifier를 못찾아가서 텍스트가 화면에 안나타나더라구요

다만 예외처리를 하려면 별도의 에러 객체도 만들어야하고, 함수도 에러를 던지는 throwing method로 변경해야 해서
지금처럼 Utils 모듈에 있는 것 보다는 SharedRepository 모듈에 별도의 Repository 클래스로 분리를 해야할 것 같네요

junu0516 replied:
넵 추가적으로 이슈 파서 해야하는 작업인듯 한데, 추후 시간될 때 진행하면 될듯합니다 ㅎㅎ

junu0516 commented on Projects/Shared/SharedUseCase/Sources/DefaultLoginTypeUseCase.swift:
```diff
@@ -0,0 +1,29 @@
+import SharedDIContainer
+import Utils
+
+public final class DefaultLoginTypeUseCase: LoginTypeUseCase {
+    @Inject(SharedDIContainer.shared) private var userDefaultsRepository: UserDefaultsRepository
+    
+    public init() { }
+    
+    public var loginType: BehaviorSubject<LoginType?> {
+        return BehaviorSubject<LoginType?>(value: userDefaultsRepository.getValue(for: .userLoginType))
+    }
+    
+    public func updateLoginType(to newLoginType: LoginType) {
+        userDefaultsRepository.updateValue(for: .userLoginType, with: newLoginType)
+    }
+}
```
이 부분은 인증 로직에 해당 UseCase 사용할 때, 로그인 정보를 UserDefaults 대신 KeyChain에서 가져오는걸로 수정해볼께요!
아무래도 인증 관련 정보는 민감하니까 KeyChain이 나을듯해서 그렇게 진행하려 하는데, 혹시 다른 방법이 괜찮다 싶으면 말씀해주세요

sanghyeok-kim replied:
네 좋습니다 이건 Keychain으로 분리하는걸로 하죠

junu0516 commented on Projects/Shared/SharedRepository/Sources/DefaultUserDefaultsRepository.swift:
```diff
@@ -0,0 +1,41 @@
+import SharedUseCase
+import Utils
+
+public final class DefaultUserDefaultsRepository: UserDefaultsRepository {
+    
+    private let userDefaults = UserDefaults.standard
+    
+    public init() { }
+    
+    public func getValue<T: Codable>(for key: UserDefaultsKeys) -> T? {
+        if let data = userDefaults.data(forKey: key.rawKey) {
+            let value = try? PropertyListDecoder().decode(T.self, from: data)
+            return value
+        }
+        
+        return userDefaults.object(forKey: key.rawKey) as? T
+    }
+    
+    public func updateValue<T: Codable>(for key: UserDefaultsKeys, with newValue: T) {
+        if let data = try? PropertyListEncoder().encode(newValue) {
+            userDefaults.set(data, forKey: key.rawKey)
+            return
+        }
+        
+        userDefaults.setValue(newValue, forKey: key.rawKey)
+    }
+    
+    public func removeValue(for key: UserDefaultsKeys) {
+        userDefaults.removeObject(forKey: key.rawKey)
+    }
```
UserDefaults에 대한 CRUD는 propertyWrapper 사용도 가능해보여요!
저는 [예시](https://medium.com/axel-springer-tech/swift-property-wrappers-user-defaults-and-unique-default-values-e8725d7154a2) 와 비슷한 방법을 종종 사용하는데, 혹시 몰라 같이 공유합니다 ㅋㅋ

sanghyeok-kim replied:
안그래도 저번 프로젝트때 UserDefaults에 대한 PropertyWrapper를 만들어 본 적이 있어서
이번에도 그렇게 구현하려했었는데,

구현하다 보니 저렇게 UserDefaults에 대한 CRUD 로직을 내부적으로 갖는 PropertyWrapper가 있으면
Repository가 하는 역할을 PropertyWrapper가 이미 하고 있는거니까, Repository가 굳이 필요 없어지더라구요.

따라서 UseCase가 UserDefaults에 대한 읽기 쓰기를
1. PropertyWrapper로 감싼 프로퍼티 형태로 갖고있을 것인지,
2. 아니면 지금 구현된 것 처럼 공통의 Repository 클래스를 통해서 할 것인지

이 두 가지 방법 중 하나를 선택하면 됐던 상황이었고

지금 제가 구현한 UserDefaults에 대한 CRUD 코드를
PropertyWrapper의 getter / setter 블록 안에 모두 구현하기에는 코드가 복잡해져서
Repository 클래스를 별도로 구현하는 것으로 결정했습니다.

sanghyeok-kim replied:
추가로 [이 코멘트](https://github.com/LifePoop/LifePoop_iOS/pull/47#discussion_r1214225070)에 대한 리팩토링을 하고 있는데,
```swift
func getValue<T: Codable>(for key: UserDefaultsKeys) -> T?
func getValue<T: RawRepresentable>(for key: UserDefaultsKeys) -> T? where T.RawValue: Codable
func updateValue<T: Codable>(for key: UserDefaultsKeys, with newValue: T)
func updateValue<T: RawRepresentable>(for key: UserDefaultsKeys, with newValue: T) where T.RawValue: Codable
func removeValue(for key: UserDefaultsKeys)
```
이렇게 메소드의 수가 꽤 많아져서, 별도 Repository 클래스 타입으로 구현하는게 맞는 것 같네요

다만 UserDefaults에 단순히 String, Int같은 원시 타입만 저장하면 되는 구조였다면
저도 간단하게 PropertyWrapper를 썼을 것 같습니다 ㅎㅎ

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingUseCase/Sources/DefaultUserSettingUseCase.swift:
```diff
@@ -0,0 +1,63 @@
+import SharedUseCase
+import Utils
+
+public final class DefaultUserSettingUseCase: UserSettingUseCase {
+    
+    @Inject(SharedDIContainer.shared) private var nicknameUseCase: NicknameUseCase
+    @Inject(SharedDIContainer.shared) private var loginTypeUseCase: LoginTypeUseCase
+    @Inject(SharedDIContainer.shared) private var autoLoginUseCase: AutoLoginUseCase
+    @Inject(SharedDIContainer.shared) private var feedVisibilityUseCase: FeedVisibilityUseCase
```
FeatureSetting 모듈에만 고유하게 존재하는 DefaultUserSettingUseCase는 결국, 필요한 SharedUseCase들의 조합으로 보면 되는건가요!!?

sanghyeok-kim replied:
네 FeatureUseCase 모듈에서는 SharedUseCase 모듈에서 필요한 UseCase만 골라 조합해서 사용하면 됩니다

junu0516 reviewed (commented):
수고 많으셨어요~! 시뮬레이터 돌려보고 궁금한점 위주로 코멘트 남겼습니다.

추가로, UITalbeViewDelegate를 별도의 객체를 만들어서 관리하셨던데 [delegate proxy](https://ios-development.tistory.com/152) 으로 해당 객체 대체하는 것도 괜찮을듯해서 제안드려봅니다..!
(대신 별도의 객체를 만들어야 하고, rx.delegate 이런 식으로 바인딩을 뷰컨트롤러안에 추가해야되서 혹여나 뷰컨트롤러를 가볍게 하시기 위함이었다면 목적에 적합하진 않겠네요)

sanghyeok-kim committed:
#43 refactor: PaddingButton, SuggestionButton의 접근 제어자 수정

sanghyeok-kim committed:
#43 refactor: RoundedPaddingLabel이 PaddingLabel을 상속받도록 변경하여 중복되는 코드 제거

sanghyeok-kim committed:
#43 refactor: UserDefaultsRepository의 get, update 메소드의 제네릭 타입을 Codable, RawRepresentable로 각각 분리

sanghyeok-kim committed:
#43 refactor: TextFile 열거형의 name 프로퍼티의 접근 제어자 변경 (internal -> public)

sanghyeok-kim committed:
#43 feat: RawRepresentable+index extension 구현

sanghyeok-kim committed:
#43 feat: 인덱싱 기능이 필요한 열거형 타입 Entity의 RawRepresentable이 Int 타입이 되도록 변경

sanghyeok-kim committed:
#43 refactor: FeedVisibility에 대해 enumerated 대신 Index 프로퍼티를 사용하여 시간복잡도 개선

sanghyeok-kim committed:
#43 fix: UserDefaultsRepository의 메소드에 전달되는 타입이 Codable과 RawRepresentable을 모두 채택한 경우 런타임 에러 발생하는 문제 수정

- 제네릭 타입이 Codable을 채택한 열거형일 경우(Codable과 RawRepresentable을 모두 채택할 경우), Codable에 대한 메소드가 실행되는데,
- 해당 열거형을 PropertyListEncoder/Decoder로 인코딩/디코딩 시도할 경우 런타임 에러가 발생함
- 따라서 PropertyListEncoder/Decoder가 아닌 JSONEncoder/Decoder를 사용하는 것으로 변경

sanghyeok-kim committed:
#43 refactor: 정적 타입으로 사용하던 BundleResourceAccessor를 별도의 Repository 타입으로 구현하여 에러 처리 및 로깅이 가능하도록 변경

sanghyeok-kim committed:
#43 refactor: FeedVisibility에 대해 CaseIterable의 extension이 아닌 Index 프로퍼티를 직접 사용하도록 변경

sanghyeok-kim committed:
#43 feat: UserDefaults에 대한 OSLog Category 추가

sanghyeok-kim committed:
#43 feat: Setting 화면에 대한 ToastMessage case 추가

sanghyeok-kim committed:
#43 feat: UserDefaultsError 생성 및 구현

sanghyeok-kim committed:
#43 refactor: UserDefaultsRepository를 RxSwift 활용 및 예외 처리하도록 변경

sanghyeok-kim committed:
#43 refactor: 변경된 UserDefaultsRepository에 맞춰 UseCase 리팩토링 및 에러 로깅 추가

sanghyeok-kim committed:
#43 refactor: 변경된 UseCase에 맞춰 ViewModel 리팩토링, State 바인딩 추가

sanghyeok-kim commented:
기존 Uils에 정적 타입으로 사용하던 BundleResourceAccessor를 별도의 Repository 타입으로 구현했고
그 과정에서 RxSwift를 활용하고 에러를 정의해 예외를 처리하였습니다.

또한 UserDefaults에 대해서도 리팩토링을 진행하여,
기존에 잘못된 값에 대해서도 nil을 반환하도록 구현되어있어 에러 디버깅이 어려운 점을 고려해
마찬가지로 Repository 타입으로 분리하고 에러를 로깅할 수 있도록 변경했습니다.
